### Printing the Python version

In [1]:
import sys

print("Python Version:", sys.version)

Python Version: 3.13.3 (main, Apr 26 2025, 11:42:11) [Clang 17.0.0 (clang-1700.0.13.3)]


### Arithemetic

In [2]:
x = 5 % 2
print("remainder = ", x)

remainder =  1


In [3]:
x = 51/15
print("quotient = ", x)

quotient =  3.4


In [4]:
x = 51//15
print("pure integer quotient = ", x)

pure integer quotient =  3


### User Console Input

In [5]:
user_input = input()
print("The user input = ", user_input)

 tobil


The user input =  tobil


### String Manipulation

In [6]:
words = ["Halloween", "Diwali", "Christmas", "New Year"]
festivals = f"First {words[0]}, then {words[1]}, then {words[2]} and finally {words[3]}"
festivals

'First Halloween, then Diwali, then Christmas and finally New Year'

In [7]:
tokens = festivals.split()
tokens

['First',
 'Halloween,',
 'then',
 'Diwali,',
 'then',
 'Christmas',
 'and',
 'finally',
 'New',
 'Year']

In [8]:
#print the characters in a String, 
sentence = "Judas the Hammer"
for c in sentence:
    print(c)

J
u
d
a
s
 
t
h
e
 
H
a
m
m
e
r


### Regex - Find all the Numbers in a String

In [9]:
"""
The Regex says 

r => raw string, i.e. the slashes are not for python but for the regex
slash d => 0-9 characters
+ => 1 or many
"""
senty = "there are 50 gold coins and in 7 bags"
import re
numbers = re.findall(r'\d+', senty)
numbers


['50', '7']

### Logical conditions and Branching

In [10]:
if len(user_input) > 100:
    print("that's a long input")
elif len(user_input) > 50:
    print("that's a medium sized input")
else:
    print("That's a short sized input", user_input)
          

That's a short sized input tobil


#### Switch Case

In [11]:
"""
Doesn't work. There is some versioning issue.
"""

"\nDoesn't work. There is some versioning issue.\n"

### Lists 

In [12]:
def age_sieve(age):
    if age < 18:
        return False
    else:
        return True
        

In [13]:
ages = [5, 7, 17,  18,  24,  32]
print("ages =", ages)
adults = filter(age_sieve, ages)
print(adults)
print(list(adults))

ages = [5, 7, 17, 18, 24, 32]
[18, 24, 32]


#### List Enumeration

In [14]:
sample_list = [22, 51, 33, 82, 99, 64, 83]

In [15]:
for index, item in enumerate(sample_list):
    print(f" index = {index}, item = {sample_list[index]}")
    print(f" item = {item}")

 index = 0, item = 22
 item = 22
 index = 1, item = 51
 item = 51
 index = 2, item = 33
 item = 33
 index = 3, item = 82
 item = 82
 index = 4, item = 99
 item = 99
 index = 5, item = 64
 item = 64
 index = 6, item = 83
 item = 83


#### List Comprehension: 

In [16]:
fruit_list = ['apple', 'banana', 'kiwi', 'mango', 'orange', 'pear']
e_fruit_list = [x for x in fruit_list if 'e' in x]
print(e_fruit_list)

['apple', 'orange', 'pear']


## Tuple to List

The core difference between a Python list and a tuple is mutability: lists are mutable (changeable), while tuples are immutable (unchangeable). 

In [17]:
some_tuple = (1, 2, 3, 4, 5)
print(type(some_tuple))
print(some_tuple)

<class 'tuple'>
(1, 2, 3, 4, 5)


In [18]:
derived_list = list(some_tuple)
print(type(derived_list))
print(derived_list)

<class 'list'>
[1, 2, 3, 4, 5]


#### Create a Java style array; fixed length and initialized to Nones


In [19]:
js_list = [None] * 5
js_list[2] = 13
print(js_list)

[None, None, 13, None, None]


#### Updating a list and returning a new list

In [20]:
base_list = ["anger", "frustration", "envy", "confusion"]
print("This will return None: ", base_list.append("healing"))
print("This will print the combibination: ", base_list +["peace"])

This will return None:  None
This will print the combibination:  ['anger', 'frustration', 'envy', 'confusion', 'healing', 'peace']


In [21]:
def custom_comp(input_str: str) -> str:
    return len(input_str)

In [22]:
# not in place update.. interesting that it removes duplicates.
sorted_list = sorted(base_list, key=custom_comp)
print(sorted_list)

['envy', 'anger', 'healing', 'confusion', 'frustration']


### List Manipulation - filter, map and Lambdas

In [23]:
def age_sieve(age: int) -> int:
    return age >= 18

In [24]:
ages = [5, 7, 12, 17, 18, 24, 32]
adults = list(filter(age_sieve, ages))
print("adults := ", adults)

adults :=  [18, 24, 32]


In [25]:
def get_length(word: str) -> int:
    return len(word)

In [26]:
fruit_lengths = list(map(get_length, ["apple", "banana", "cherry", "durien"]))
print("fruit lengths: ", fruit_lengths)

fruit lengths:  [5, 6, 6, 6]


Using an anonymous lambda below:

In [27]:
fruit_len_iterable = map(lambda x: len(x), ['apple', 'banana', 'cherry'])
list(fruit_len_iterable)

[5, 6, 6]

## Set

Set has unique items, but is mutable.
An empty curly braces literal is treated as a dict, if you put items, it is treated as a set.

In [28]:
braces_literal = {}
type(braces_literal)

dict

In [29]:
braces_literal = {1}
type(braces_literal)

set

Sets are Mutable

In [30]:
first_set = {1, 2, 3}
first_set.add(1)
print(first_set)
print(first_set.add(4))
first_set.remove(3)
print(first_set)
print(2 in first_set)

{1, 2, 3}
None
{1, 2, 4}
True


Using the set() constructor

In [31]:
second_set = set()
print(second_set)

set()


In [32]:
second_set.add("hammer")
print(second_set)

{'hammer'}


In [33]:
second_set.remove("hammer")
print(second_set)

set()


A frozen set is an immutatble set - once created, you cannot add or remove items

In [34]:
third_set = frozenset({1, 5, 99})
print(third_set)
print(type(third_set))

frozenset({1, 99, 5})
<class 'frozenset'>


Nested Sets - You can only add a frozen set inside a set

In [35]:
set_outer = set()
set_inner = set()
set_inner.add(2)
# set_outer.add(set_inner) - wil not work, TypeError: unhashable type: 'set'
frozen_inner = frozenset(set_inner)
set_outer.add(frozen_inner)

Appending in place vs using the pipe operator

In [36]:
print("This operation has no return type, it is an in-place append", set_inner.add(3))

This operation has no return type, it is an in-place append None


In [37]:
print("This operation returns a new set", set_inner | {4} )

This operation returns a new set {2, 3, 4}


# Dict

In [39]:
first_dict = {}
first_dict.update({1: "a"})
print(first_dict)
print(first_dict[1])

{1: 'a'}
a


In [45]:
first_dict = {1: "one", 2: "two", 3: "three"}

for key in first_dict.keys():
    print(key, " = ", first_dict[key])

for key, value in first_dict.items():
    print(key, " = ", value)

1  =  one
2  =  two
3  =  three
1  =  one
2  =  two
3  =  three


## Default Dict

Transform an array of objects into an dict of arrays

In [46]:
transaction_list = [
    {'date': '2023-11-01', 'product': 'A', 'amount': 100, 'customer': 'Alice'},
    {'date': '2023-11-02', 'product': 'B', 'amount': 150, 'customer': 'Bob'},
    {'date': '2023-11-03', 'product': 'A', 'amount': 200, 'customer': 'Charlie'},
    {'date': '2023-11-04', 'product': 'C', 'amount': 50, 'customer': 'David'},
    {'date': '2023-11-05', 'product': 'B', 'amount': 75, 'customer': 'Eve'},
]

In [51]:
trans_map = {}
for trans in transaction_list:
    product = trans['product']
    if product not in trans_map:
        trans_map.update({product: []})
    trans_map[product].append(trans)
print(trans_map)    

{'A': [{'date': '2023-11-01', 'product': 'A', 'amount': 100, 'customer': 'Alice'}, {'date': '2023-11-03', 'product': 'A', 'amount': 200, 'customer': 'Charlie'}], 'B': [{'date': '2023-11-02', 'product': 'B', 'amount': 150, 'customer': 'Bob'}, {'date': '2023-11-05', 'product': 'B', 'amount': 75, 'customer': 'Eve'}], 'C': [{'date': '2023-11-04', 'product': 'C', 'amount': 50, 'customer': 'David'}]}


Do this better with a defaultdict

In [57]:
from collections import defaultdict
trans_map = defaultdict(list)
for trans in transaction_list:
    product = trans['product']
    trans_map[product].append(trans)
print(trans_map)

defaultdict(<class 'list'>, {'A': [{'date': '2023-11-01', 'product': 'A', 'amount': 100, 'customer': 'Alice'}, {'date': '2023-11-03', 'product': 'A', 'amount': 200, 'customer': 'Charlie'}], 'B': [{'date': '2023-11-02', 'product': 'B', 'amount': 150, 'customer': 'Bob'}, {'date': '2023-11-05', 'product': 'B', 'amount': 75, 'customer': 'Eve'}], 'C': [{'date': '2023-11-04', 'product': 'C', 'amount': 50, 'customer': 'David'}]})


# Tuple

Tuples - 

1. Immutable collections - ordered as well
2. Multiple return values from methods
3. As shown above, they are iterables which can be converted to a list.

In [1]:
sample_tuple = (7, 4, 5, 6, 8, 10)
for i in sample_tuple:
    print(i)

7
4
5
6
8
10


# Global Interpreter Lock - GIL

1. GIL is the Global Interpreter Lock in Python, it is the default behavior and cannot be overridden.
2. GIL makes sure that only one thread can access the Python Interpreter at a time.
3. As an analogy in Java, even if you have code that is multi threaded, if the machine has a single core, only one thread runs at a time. In Python, the GIL makes even a multi-core machine behave like a single core machine.
4. So, multi threading in Python is useful only if the threads do remote or I/O operations.
5. Alternately, you can do multi-processing. Then each process will have its own Interpreter and GIL
6. There is talk about diminishing or eliminating the GIL. Starting with Python 3.13, CPython introduced an experimental free-threaded build where the GIL can be disabled.

# Functions

In [1]:
def first_function():
    print("Hello from inside a function")

first_function()    

Hello from inside a function


In [2]:
def second_function(param1: str):
    print(f"This was sent as input to the function {param1}")

second_function("Sample Param")    

This was sent as input to the function Sample Param


In [3]:
# Using Var args
def third_function(*params):
    for i, param in enumerate(params):
        print(f"parameter numer {i} is {param}")



In [4]:
third_function(100, 200, 300)

parameter numer 0 is 100
parameter numer 1 is 200
parameter numer 2 is 300


In [6]:
third_function([100, 200, 300])

parameter numer 0 is [100, 200, 300]


In [7]:
third_function("This is a string")        

parameter numer 0 is This is a string


### Keyword Args or Kwargs

In [9]:
# There is nothing particularly different in the invocation, not even the default values, it is just the invocation
def fourth_function(param1, param2):
    print(f"param1 = {param1}, param2 = {param2}")
    

In [11]:
# fourth_function(param2="hello", 2) Does not work - positional argument follows keyword argument
fourth_function(3, param2="hello")

param1 = 3, param2 = hello


In [12]:
fourth_function(param2="Hello", param1="Hi") # works, the order is maintained despite being jumbled in invocation

param1 = Hi, param2 = Hello


In [13]:
def fifth_function(*params, **key_params):
    print("Positional params = ", params)
    print("Keyword params = ", key_params)

In [14]:
fifth_function(1, 2, 3)

Positional params =  (1, 2, 3)
Keyword params =  {}


In [15]:
fifth_function(1, 2, gauge=12, barrel=2)

Positional params =  (1, 2)
Keyword params =  {'gauge': 12, 'barrel': 2}


# Typing Hints

With Python 3.9, it looks there are 2 ways to provide type hints

https://stackoverflow.com/questions/39458193/using-list-tuple-etc-from-typing-vs-directly-referring-type-as-list-tuple-etc 

In [16]:
# using the lower case list as the type hint
def square_list(in_list: list[int]) -> list[int]:
    return [x*x for x in in_list]

In [18]:
print(square_list([3, 5, 7]))

[9, 25, 49]


In [19]:
from typing import List
def cube_list(in_list: list[int]) -> list[int]:
    return [x*x*x for x in in_list]
    

In [21]:
print(cube_list([2, 3, 5]))

[8, 27, 125]


Type aliases are also available
https://docs.python.org/3/library/typing.html#type-aliases

you can use the "type" keyword to create a new type; but it is a type alias—not a genuinely new type

In [22]:
Vector = List[float]

In [25]:
def scale(scalar: float, vector: Vector) -> Vector:
    return [scalar * element for element in vector]

In [26]:
scaled_vector = scale(2.0, [1, -4.2, 5.4])
print(scaled_vector)

[2.0, -8.4, 10.8]
